### DISCUSSION TITLE
Better Monorepo support through package rule template rendering

### DISCUSSION THREAD

**Discussion by ricochet1k** [renovatebot/renovate#23436] [Suggest an Idea]
### Tell us more.

I've been looking into a way to modify Renovate to better support our monorepo layout. Basically, we have several repos with a very wide directory structure. Some of these are leaf projects with only dependencies, some of them are a big collection of libraries that are depended on.

So there are two basic concerns:
- We'd like a PRs created by Renovate to be grouped by the subdirectory the updates are contained in. `groupName` is the obvious way to do this, but there's no way currently to generate the `groupName` with a template/regex match, and I don't want to create or maintain the list of several hundred subdirectories at varying depths.
- We'd also like Terragrunt dependencies to be matched by the subdirectory or tag prefix of the tag in the dependency reference. The versioning `compatibility` match group almost does this, but it doesn't get used for dependency grouping or in the PR/commit name, and again, there's no way currently to alter those with a template/regex match in package rules.

My proposal is this: 
- Alter the `match*` package rules to return the regex match groups when they match (this isn't applicable to all of the matchers obviously)
- Add some way to render templates with access to those matched groups to package rules when the rule matches. I implemented this with just a `templates` object, where every value gets rendered and used as the value to merge into the config. This might be a little too flexible, but it's easy and gets the job done. It's also entirely possible that this opens up solutions to other problems.

I've actually already made these changes here: https://github.com/ricochet1k/renovate/tree/package-rule-templates
Tests aren't quite fixed up yet, but I could fix them if needed.

With these changes I'm able to use a config similar to this to get the job done:
```js
module.exports = {
  "packageRules": [
    {
      "matchDepNames": ["github.com/blah/some-utils"],
      "matchCurrentValue": "/^((?<depName>.*)-v|v*)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$/",
      "matchFiles": "/^(?<groupName>([^/]+/)+)terragrunt.hcl/",
      "templates": {
        "groupName": "{{{groupName}}} {{{depName}}}",
        "depName": "some-utils#{{{depName}}}",
        commitMessageTopic: '{{{groupName}}} {{{depName}}}',
      },
      "versioning": "regex:^((?<compatibility>.*)-v|v*)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$",
    },
  ],

  "group": {
    branchTopic: '{{{groupSlug}}}',
    commitMessageTopic: '{{{groupName}}} {{{depName}}}',
  },
}
```

**Comment by rarkins** [comment]
Try additionalBranchPrefix with templated parentDir - search the discussions here for past examples
